src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -303,20 +257,23 @@ public override int GetHashCode()
         /// <param name="text">The <see cref="string"/>to compare.</param>
         /// <param name="comparisonType">One of the enumeration values that specifies the rules to use in the comparison.</param>
         /// <returns><code>true</code> if <paramref name="text"/> matches the beginning of this <see cref="StringSegment"/>; otherwise, <code>false</code>.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public bool StartsWith(string text, StringComparison comparisonType)
         {
             if (text == null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
             }
 
-            var textLength = text.Length;
-            if (!HasValue || Length < textLength)
+            bool result = false;
+            int textLength = text.Length;
```

Comment:
Similar effect to your earlier `Equals` change with advantage to just using `text.Length` directly?

Reply:
Unfortunately not. 

This PR:
```asm
G_M15035_IG03:
       33C0                 xor      eax, eax
       488975F8             mov      gword ptr [rbp-08H], rsi
       448B4608             mov      r8d, dword ptr [rsi+8]
       48833F00             cmp      gword ptr [rdi], 0
       7428                 je       SHORT G_M15035_IG04
       4439470C             cmp      dword ptr [rdi+12], r8d
       7C22                 jl       SHORT G_M15035_IG04
       488B17               mov      rdx, gword ptr [rdi]
       8B7708               mov      esi, dword ptr [rdi+8]
       488BFA               mov      rdi, rdx
       488B55F8             mov      rdx, gword ptr [rbp-08H]
       33C9                 xor      ecx, ecx
       41B904000000         mov      r9d, 4
       E8C1FD28FF           call     System.String:Compare(ref,int,ref,int,int,int):int
```

Without local `textLength`:
```asm
G_M15035_IG03:
       33C0                 xor      eax, eax
       48833F00             cmp      gword ptr [rdi], 0
       7437                 je       SHORT G_M15035_IG04
       448B470C             mov      r8d, dword ptr [rdi+12]
       488975F8             mov      gword ptr [rbp-08H], rsi
       443B4608             cmp      r8d, dword ptr [rsi+8]
       7C29                 jl       SHORT G_M15035_IG04
       488B17               mov      rdx, gword ptr [rdi]
       8B7708               mov      esi, dword ptr [rdi+8]
       488B4DF8             mov      rcx, gword ptr [rbp-08H]
       448B4108             mov      r8d, dword ptr [rcx+8]
       488BFA               mov      rdi, rdx
       488BD1               mov      rdx, rcx
       33C9                 xor      ecx, ecx
       41B904000000         mov      r9d, 4
       E8BAFD28FF           call     System.String:Compare(ref,int,ref,int,int,int):int
```

So with the local there's a better reuse of the values for String:Compare.